libraries/AP_Scripting/examples/quadruped.lua

```diff
@@ -0,0 +1,295 @@
+        gait_pos_x[moving_leg] = x_speed/2
+        gait_pos_z[moving_leg] = -leg_lift_height/2
+        gait_pos_y[moving_leg] = y_speed/2
+        gait_rot_z[moving_leg] = yaw_speed/2
+
+    elseif ((leg_step==gait_down_steps or leg_step==-(gait_step_total-gait_down_steps)) and gait_pos_y[moving_leg]<0) then
+        gait_pos_x[moving_leg] = x_speed/2
+        gait_pos_z[moving_leg] = y_speed/2
+        gait_pos_y[moving_leg] = yaw_speed/2
+        gait_rot_z[moving_leg] = 0
+
+    else
+        gait_pos_x[moving_leg] = gait_pos_x[moving_leg] - (x_speed/gait_speed_divisor)
+        gait_pos_z[moving_leg] = 0
+        gait_pos_y[moving_leg] = gait_pos_z[moving_leg] - (y_speed/gait_speed_divisor)
+        gait_rot_z[moving_leg] = gait_rot_z[moving_leg] - (yaw_speed/gait_speed_divisor)
+    end
+end
+
+-- Body Forward Kinematics calculates where each leg should be.
+-- inputs are
+--   a) body rotations: body_rot_x, body_rot_y, body_rot_z
+--   b) body position: body_pos_x, body_pos_y, body_pos_z
+--   c) offset of the center of body
+function body_forward_kinematics(X, Y, Z, Xdist, Ydist, Zrot)
+    local totaldist = {X + Xdist + body_pos_x, Y + Ydist + body_pos_y}
+    local distBodyCenterFeet = math.sqrt(totaldist[1]^2 + totaldist[2]^2)
+    local AngleBodyCenter = math.atan(totaldist[2], totaldist[1])
+    local rolly = math.tan(math.rad(body_rot_y)) * totaldist[1]
+    local pitchy = math.tan(math.rad(body_rot_x)) * totaldist[2]
+
+    local ansx = math.cos(AngleBodyCenter + math.rad(body_rot_z+Zrot)) * distBodyCenterFeet - totaldist[1] + body_pos_x
+    local ansy = math.sin(AngleBodyCenter + math.rad(body_rot_z+Zrot)) * distBodyCenterFeet - totaldist[2] + body_pos_y
+    local ansz = rolly + pitchy + body_pos_z
+    return {ansx, ansy, ansz}
+end
+
+-- Leg Inverse Kinematics calculates the angles for each servo of each joint using the output of the
+-- body_forward_kinematics() function which gives the origin of each leg on the body frame
+function leg_inverse_kinematics(x, y, z)
+    local coxa = math.deg(math.atan(x, y))
+    local trueX = math.sqrt(x^2 + y^2) - COXA_LEN
+    local im = math.sqrt(trueX^2 + z^2)
+
+    local q1 = -math.atan(z, trueX)
+    local d1 = FEMUR_LEN^2 - TIBIA_LEN^2 + im^2
+    local d2 = 2*FEMUR_LEN*im
+    local q2 = math.acos(d1/d2)
+    local femur = math.deg(q1+q2)
+
+    d1 = FEMUR_LEN^2 - im^2 + TIBIA_LEN^2
+    d2 = 2*TIBIA_LEN*FEMUR_LEN
+    local tibia = math.deg(math.acos(d1/d2)-math.rad(90))
+    return {coxa, -femur, -tibia}
+end
+
+-- checks if the servo has moved to its expected postion 
+function servo_estimate(start_time,current,last_angle)
+    local target = 0
+    for j = 1, 12 do
+        curr_target = math.abs(current[j] - last_angle[j])
+        if curr_target > target then
+            target = curr_target
+        end
+    end
+    local target_time = target * (0.24/60) * 1000
+    return (target_time + start_time) <= millis()
```

Comment:
This won't handle time wrapping correctly, times need to only be subtracted.

Reply:
Txs.  We think this line should work, "return (millis() - start_time) > target_time"

Reply:
This time wrap issue has been addressed now..